genienlp/data_utils/bootleg.py

```diff
@@ -0,0 +1,381 @@
+import os
+import ujson
+import numpy as np
+import logging
+
+from .database_utils import is_banned
+
+from bootleg.extract_mentions import extract_mentions
+from bootleg.utils.parser_utils import get_full_config
+from bootleg import run
+
+
+logger = logging.getLogger(__name__)
+
+
+def reverse_bisect_left(a, x, lo=None, hi=None):
+    """Find item x in list a, and keep it reverse-sorted assuming a
+    is reverse-sorted.
+    """
+    if lo is None:
+        lo = 0
+    if hi is None:
+        hi = len(a)
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if x > a[mid]:
+            hi = mid
+        else:
+            lo = mid + 1
+    return lo
+
+
+def post_process_bootleg_types(qid, type, title, almond_domains):
+    # TODO if training on multiple domains (in one run) these mapping should be modified
+    # e.g. song is mapped to book which is not correct if training on music domain too
+    for domain in almond_domains:
```

Comment:
When you train a multidomain Mario model, what domains do you pass?
This is ok for now because the PR needs to land, but you need to refactor this ASAP. Make some JSON file that Genie can generate with the information you need.

Reply:
Only Spotify. Yeah, domain-specific codes will be gone (or at least reduced) in the next PR.